R/coerce.R

```diff
@@ -0,0 +1,180 @@
+#'                     "barcode", "in_tissue", "array_row", "array_col",
+#'                     "pxl_row_in_fullres", "pxl_col_in_fullres"))
+#' 
+#' (spe <- as(sce, "SpatialExperiment"))
+#' ## as method
+#' int_colData(sce)$spatialData <- DataFrame(xyz[,c(1:4)])
+#' int_colData(sce)$spatialCoords <- as.matrix(xyz[,c(5,6)])
+#' ## Coercing an sce without imgData
+#' (spe <- as(sce, "SpatialExperiment"))
+#' ## Coercing an sce with imgData
+#' img <- readImgData(
+#'     path = file.path(dir, "spatial"),
+#'     sample_id = "sample01")
+#' int_colData(sce)$imgData <- img
+#' (spe <- as(sce, "SpatialExperiment"))
+#' 
+#' ## toSpatialExperiment method
+#' (spe <- toSpatialExperiment(sce,
+#'     imgData = img,
+#'     spatialData = DataFrame(xyz), 
+#'     spatialCoordsNames = c("pxl_col_in_fullres", "pxl_row_in_fullres"),
+#'     sample_id = "sample01"))
+NULL
+
+
+setAs(
+    from="SingleCellExperiment", 
+    to="SpatialExperiment", 
+    function(from) {
+        spatialCoords <- spatialData <- imgData <- NULL
+        sample_id <- "sample01"
+        if ("spatialCoords" %in% names(int_colData(from))) {
+            spatialCoords <- int_colData(from)$spatialCoords
+        }
+        if ("spatialData" %in% names(int_colData(from))) {
+            spatialData <- int_colData(from)$spatialData
+        }
+        if ("imgData" %in% names(int_colData(from))) {
+            imgData <- int_colData(from)$imgData
+        }
+        if ("sample_id" %in% colnames(colData(from))) {
+            sample_id <- unique(colData(from)$sample_id)
+        }
+        spe <- .sce_to_spe(from, 
+            sample_id=sample_id,
+            spatialData=spatialData,
+            spatialCoords=spatialCoords,
+            imgData=imgData)
+        return(spe)
+    }
+)
+
+#' @export
+toSpatialExperiment <- function(sce,
+                                sample_id="sample01", 
+                                spatialDataNames=NULL,
+                                spatialCoordsNames=NULL,
+                                spatialData=NULL,
+                                spatialCoords=NULL,
+                                scaleFactors=1,
+                                imageSources=NULL,
+                                image_id=NULL,
+                                loadImage=TRUE,
+                                imgData=NULL) {
+    
+    stopifnot(is(sce, "SingleCellExperiment"))
+    
+    ## giving priority to passed arguments
+    if (all(is.null(spatialCoords), is.null(spatialCoordsNames),
+        "spatialCoords" %in% names(int_colData(sce))) ) {
+        spatialCoords <- int_colData(sce)$spatialCoords
+        spatialCoordsNames <- names(spatialCoords)
+    }
+    if (all(is.null(spatialData), is.null(spatialDataNames),
+        "spatialData" %in% names(int_colData(sce))) ) {
+        spatialData <- int_colData(sce)$spatialData
+    }
+    if (all( is.null(imgData),
+             "imgData" %in% names(int_colData(sce))) ) {
```

Comment:
see comment above: we don't need the `"xxx" in names(...)` but can just assign `xxx <- int_colData(sce)$xxx` if the corresponding value isn't provided as an argument.

Reply:
in this case this is true only for the imgData, because in the other cases we have the xxxNames

Reply:
No, what I mean is that you can just put
```
    if (is.null(spatialCoords) && is.null(spatialCoordsNames)) {
        spatialCoords <- int_colData(sce)$spatialCoords
        spatialCoordsNames <- names(spatialCoords)
    }
    if (is.null(spatialData) && is.null(spatialDataNames)) {
        spatialData <- int_colData(sce)$spatialData
    }
    if (all(is.null(imgData))) {
        imgData <- int_colData(sce)$imgData
    }
```

Reply:
done